### PR TITLE
Initial Bigquery Client Support 

### PR DESCRIPTION
**Pull request by chekolyn**
Basic implementation for BigQuery client support: https://github.com/beekeeper-studio/beekeeper-studio/issues/83 

Only Service Account authentication is working; since Google Authentication can be implemented in numerous ways.

### CODE REVIEW

**Comment by rathboma** [comment]
Whaaaaat!!? This is awesome. We'll get this reviewed ASAP. Thanks for the amazing PR!

File: apps/studio/src/lib/db/clients/bigquery.js [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+}
+
+
+function prettyJSON(obj) {
+  logger().debug(JSON.stringify(obj, null, 2))
+}
+
+
+async function listTablesOrViews(client, db, type) {
+  // Lists all tables or views in the dataset
```
Comment (not-night-but) [inline apps/studio/src/lib/db/clients/bigquery.js:191]: This throws an error if there is no dataset currently selected (connecting without a default dataset for instance.

**Review by not-night-but — CHANGES_REQUESTED** [review]
@chekolyn Hey! Sorry it took a bit to get back to you on this, we've had a lot going on.

I haven't gone super in depth into the code yet, but some initial issues I'm finding when messing around with this locally:

- [x] It seems the implementation isn't checking that the connection was successful before proceeding to the core interface (ie query editor, tables list). So I can actually 'connect' to a bigquery database even if I don't have one running.
- [x] I'm also having trouble with just running queries in general, but I suspect this has to do with the above issue, so I'm not even sure if my connection is valid. I will continue messing around with it
- [x] Add database doesn't currently work `The database could not be created: Cannot read properties of undefined (reading 'apply')`
- [ ] Once we get the basic functionality working we should probably write some integration tests similar to how the other databases are implemented if possible. Would help with making sure everything is functioning correctly. 

This is a great start! Looks like all the foundations are there, we just need to figure out what is going on with the connection interface issue and then I'm sure the rest of the issues will become much easier to debug.

I did pull down the PR and fix the build/lint issues that are preventing the workflows from running properly.

I will go through your implementation soon and give you feedback on that, but I think these issues are more pressing than code style

**Comment by chekolyn** [comment]
@not-night-but thanks for the feedback will get working on the above.

File: apps/studio/src/lib/db/clients/bigquery.js (hunk 1 above)
Comment (chekolyn) [reply apps/studio/src/lib/db/clients/bigquery.js:191]: I have found the bug for this; this line is not needed in the implementation. 
```server.config.client = client // save the client for later use```

**Comment by not-night-but** [comment]
@chekolyn yay! I'll take a look today!

> I was hopping you can provide more insight regarding the built/lint issues; not sure were to look regarding that.

Sorry what build/lint issues are you experiencing? The workflows seem to have passed? When developing if you want to check linting you should be able to run `yarn run all:lint` and that will lint/autofix any issues

File: apps/studio/src/migration/dev-1.js [hunk 2]
```diff
@@ -44,6 +44,14 @@ export default {
...
       defaultDatabase: 'sakila'
     },
+    {
+      name: "[DEV] Docker Bigquery",
+      connectionType: 'bigquery',
+      port: 9050,
+      host: 'localhost',
+      projectId: 'bks',
+      defaultDatabase: 'world',
+    },
```
Comment (not-night-but) [inline apps/studio/src/migration/dev-1.js:54]: This should be in it's own migration, look at dev-3.js on how to do that. The reason is, for existing development builds this won't run.

**Review by not-night-but — CHANGES_REQUESTED** [review]
@chekolyn Great work so far! I still have to go through a bunch of the code, but right now there is one big issue: we need a dialect data file for this. You can look in `shared/src/lib/dialects/` for examples of this, and that is where it should go. 
This file provides a lot of functionality for the app (you may notice the query editor doesn't work as well as the other dialects right now, this is the reason). It seems that it is just defaulting to the SQLite dialect data, which is not what we want:
image

I'll continue to go through the code and add feedback. If you could resolve conversations that you have addressed that would be great.

I merged master into your branch so there are no conflicts at the moment. You're welcome to ping me if you need any clarification on things.

Thanks again!
Day

**Comment by chekolyn** [comment]
Hola @not-night-but ,

I've been looking at the dialect code and it's going to take me a little while to figure it out. This part of the code also uses the knex module, the good news is that the sql-formatter does have a dialect for **bigquery** out of the box. 
 
> @chekolyn Great work so far! I still have to go through a bunch of the code, but right now there is one big issue: we need a dialect data file for this. You can look in `shared/src/lib/dialects/` for examples of this, and that is where it should go. This file provides a lot of functionality for the app (you may notice the query editor doesn't work as well as the other dialects right now, this is the reason). It seems that it is just defaulting to the SQLite dialect data, which is not what we want: image

**Comment by not-night-but** [comment]
@chekolyn alrighty. I'll look into that knex extension you linked, but it looks like it's not maintained so we might have to fork it 🤷

**Comment by not-night-but** [comment]
@chekolyn Hey! So just to keep you updated, we are gonna get this merged pretty soon. This will be an "alpha" release of the feature though, as there are probably still some things to clean up. Great work though! 

Thank you so much for contributing,
Day

**Comment by chekolyn** [comment]
Hola @not-night-but,

That's great news! Please let me know if any additional help is needed.

**Comment by rathboma** [comment]
@not-night-but thought we were merging this then fixing? Lets get it merged.

**Comment by not-night-but** [comment]
Yeah sorry, I'm working on another branch, just haven't clicked the button yet.

**Review by not-night-but — APPROVED** [review]

**Pull request merged**